miniapp/brunel/brunel_miniapp.cpp

```diff
@@ -0,0 +1,326 @@
+#include <communication/communicator.hpp>
+#include <communication/global_policy.hpp>
+#include <cell.hpp>
+#include <io/exporter_spike_file.hpp>
+#include <lif_cell_description.hpp>
+#include <lif_cell_group.hpp>
+#include <model.hpp>
+#include <profiling/profiler.hpp>
+#include <profiling/meter_manager.hpp>
+#include <pss_cell_description.hpp>
+#include <pss_cell_group.hpp>
+#include <recipe.hpp>
+#include <set>
+#include <threading/threading.hpp>
+#include <util/config.hpp>
+#include <util/debug.hpp>
+#include <util/ioutil.hpp>
+#include <util/nop.hpp>
+#include <vector>
+
+#include "io.hpp"
+
+using namespace nest::mc;
+
+using global_policy = communication::global_policy;
+using file_export_type = io::exporter_spike_file<global_policy>;
+void banner();
+using communicator_type = communication::communicator<communication::global_policy>;
+
+// Samples m unique values in interval [start, end) - gid.
+// We exclude gid because we don't want self-loops.
+std::vector<int> sample_subset(int gid, int start, int end, int m) {
+    std::set<int> s;
+    
+    std::mt19937 gen(gid + 42);
+    std::uniform_int_distribution<int> dis(start, end - 1);
+    while (s.size() < m) {
+        auto val = dis(gen);
+        if (val != gid) {
+            s.insert(val);
+        }
+    }
+    
+    return {s.begin(), s.end()};
+}
+
+/*
+ Brunel networks consists of nexc excitatory LIF neurons, ninh inhibitory LIF neurons and
+ nexc Poisson neurons. Each neuron in the network receives in_degree_prop * nexc excitatory connections
+ chosen randomly, in_degree_prop * ninh inhibitory connections and in_degree_prop * nexc Poisson connections.
+ All the connections have the same delay. The strenght of excitatory and poisson connections is given by
+ parameter weight, whereas the strength of inhibitory connections is rel_inh_strength * weight.
+ Poisson neurons all spike independently with mean frequency poiss_rate [kHz].
+ Because of the refractory period, the activity is mostly driven by poisson neurons and
+ recurrent connections have just a small effect.
+ */
+class brunel_recipe: public recipe {
+public:
+    brunel_recipe(cell_size_type nexc, cell_size_type ninh, double in_degree_prop, float weight, float delay, float rel_inh_strength, double poiss_rate):
+    ncells_exc_(nexc), ncells_inh_(ninh), delay_(delay), rate_(poiss_rate)
+    {
+        // Make sure that in_degree_prop in the interval (0, 1]
+        if (in_degree_prop <= 0.0 || in_degree_prop > 1.0) {
+            std::out_of_range("The proportion of incoming connections should be in the interval (0, 1].");
+        }
+        
+        // Set up the parameters.
+        ncells_ext_ = nexc;
+        weight_exc_ = weight;
+        weight_inh_ = -1.0 * rel_inh_strength * weight_exc_;
+        weight_ext_ = weight;
+        in_degree_exc_ = (int) (in_degree_prop * nexc);
+        in_degree_inh_ = (int) (in_degree_prop * ninh);
+        in_degree_ext_ = (int) (in_degree_prop * ncells_ext_);
```

Comment:
casting to `int` will always round-down.
Would `std::round()` be more appropriate here?
Should we check that the sum of `in_degree_{exc,inh,ext}_ == in_degree_prop * num_cells()`?